File: bindings/julia/src/SafeCrypto.jl
# SPDX-License-Identifier: PMPL-1.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeCrypto

Cryptographic safety operations with constant-time guarantees.
"""
module SafeCrypto

export constant_time_compare, secure_zero!

"""
    constant_time_compare(a::AbstractVector{UInt8}, b::AbstractVector{UInt8}) -> Bool

Compare two byte arrays in constant time to prevent timing attacks.
"""
function constant_time_compare(a::AbstractVector{UInt8}, b::AbstractVector{UInt8})::Bool
    length(a) != length(b) && return false
    isempty(a) && return true

    result = UInt8(0)
    for i in eachindex(a)
        result |= a[i] ⊻ b[i]
    end
    result == 0
end

"""
    constant_time_compare(a::AbstractString, b::AbstractString) -> Bool

Compare two strings in constant time to prevent timing attacks.
"""
function constant_time_compare(a::AbstractString, b::AbstractString)::Bool
    constant_time_compare(Vector{UInt8}(a), Vector{UInt8}(b))
end

"""
    secure_zero!(data::AbstractVector{UInt8}) -> Nothing

Securely zero out a byte array to prevent data leakage.
"""
function secure_zero!(data::AbstractVector{UInt8})::Nothing
    for i in eachindex(data)
        data[i] = 0x00
    end
    # Prevent compiler from optimizing away the zeroing
    @assert all(==(0x00), data)
    nothing
end

"""
    secure_zero!(s::Base.SecretBuffer) -> Nothing

Securely zero a SecretBuffer (Julia's built-in secure string type).
"""
function secure_zero!(s::Base.SecretBuffer)::Nothing
    Base.shred!(s)
    nothing
end

end # module

File: bindings/ocaml/lib/safe_crypto.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Cryptographic safety operations with constant-time guarantees. *)

let constant_time_compare a b =
  if String.length a <> String.length b then false
  else if String.length a = 0 then true
  else
    let result = ref 0 in
    for i = 0 to String.length a - 1 do
      result := !result lor (Char.code a.[i] lxor Char.code b.[i])
    done;
    !result = 0

let secure_zero length =
  String.make length '\000'

let random_bytes n =
  let buf = Bytes.create n in
  let ic = open_in_bin "/dev/urandom" in
  really_input ic buf 0 n;
  close_in ic;
  Bytes.to_string buf

let random_hex n =
  let bytes = random_bytes n in
  let buf = Buffer.create (n * 2) in
  String.iter (fun c ->
    Printf.bprintf buf "%02x" (Char.code c)
  ) bytes;
  Buffer.contents buf
